### ISSUE TITLE
async conflict with ansible_user and --become flag

### ISSUE THREAD

**Issue by bbyhuy** [ansible/ansible#14568]
##### Issue Type:
- Bug Report
##### Ansible Version:

```
ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

host_key_checking = false
##### Environment:

Ubuntu 14.04
##### Summary:

I think it is related to:
[https://github.com/ansible/ansible/issues/13965]

My host file specifies the `ansible_user=ubuntu`
My command: `ansible-playbook site.yml -i myhost -b -v`

My playbook essentially runs a shell script that executions a compiled program and writes out a file using `async`.

What I think is happening, as mentioned in the linked bug above, is that, async_task is being written to `~root/.ansible_async/*` and not `~ubuntu/.ansible_async/*`

So while I can see the program compiled correctly by accessing the node and seeing the output file. Ansible will throw up the `"msg": "The async task did not return valid JSON: No JSON object could be decoded"` 

I tried adding `become: yes become_user: ubuntu` but that gave me some permission errors for that particular task. Which i worked around by throwing `sudo` in front of the shell command. 
##### Steps To Reproduce:
1. Simple playbook that requires sudo, apt-get or something while also using async to writing to a protected directory (dd?).
2. Set ansible_user in the host file (anything but root)
3. execute with `ansible-playbook site.yml -i myhost -b -v`
##### Expected Results:

```
2016-02-18 13:48:34,356 p=32700 u=root |  PLAY RECAP *********************************************************************
2016-02-18 13:48:34,357 p=32700 u=root |  10.115.23.218              : ok=42   changed=25   unreachable=0    failed=0
```
##### Actual Results:

```
2016-02-18 13:38:04,760 p=29753 u=root |  ^[[0;31mfatal: [10.115.23.218]: FAILED! => {"changed": false, "failed": true, "msg": "The async task did not return valid JSON: No JSON object could be decoded"}^[[0m
2016-02-18 13:38:04,762 p=29753 u=root |  PLAY RECAP *********************************************************************
2016-02-18 13:38:04,762 p=29753 u=root |  10.115.23.218              : ok=18   changed=11   unreachable=0    failed=1
```

**Comment by pshchelo** [comment]
I was experiencing the same issue and found following reason in my particular case (may be someone find it useful):

For my use-case I'm using a very minimalistic bootstrap image based on Ubuntu 14.04 built with diskimage-builder (and devuser element, passwd-less sudo user "ansible").

I was still getting "The async task did not return valid JSON" error although I checked and in my ansible version (2.0.1.0) the bug [#13965] is fixed. After applying a patch from [the ML](https://groups.google.com/forum/#!topic/ansible-devel/9WLR5wpguH8) to see what was returned instead of valid JSON I found a following string prepended to valid JSON output:

sudo: unable to resolve host ubuntu\r\n

As it turned out, the image I was working with was completely missing "/etc/hosts" file. Thus, even when logged in to the box, on every command involving "sudo" this string was printed in the command line, and obviousely passed to ansible output too (the command itself succeeds).

Fixing the hosts file (`sudo echo "127.0.0.1 localhost ubuntu" > /etc/hosts`) resolved the issue, and become+async is working now.

Bottom line - it would be nice to have possibility to get the offending string verbatim in case of failed JSON parsing, may be configurable with --debug/-v or something like that.

**Comment by ssbarnea** [comment]
I have the impression that I am facing the same bug or one that is closely related to this one. Ansible 2.1.0.0, error `The async task did not return valid JSON: No JSON object could be decoded`

```
- block:
  - name: "Reboot system if required"
    command: shutdown -r now 'Rebooting to complete system upgrade'
    args:
      removes: /var/run/reboot-required
    poll: 10
    async: 300
    ignore_errors: yes
  become: yes
  become_user: root
```

**Comment by mattclay** [comment]
PR #16534 been merged and may fix your issue. The problem may go away entirely with this fix. If not, the error message "The async task did not return valid JSON" you reported should be replaced with a more informative error, allowing further troubleshooting.

Unless further tests reveal major bugs, this fix will be included in the upcoming 2.1.1 release. If you'd like to test and confirm the fix before then, please follow the instructions on [running from source](http://docs.ansible.com/ansible/intro_installation.html#running-from-source). Be sure to use `-b stable-2.1` when you clone so you are testing the stable-2.1 branch.

If we don't hear back before the 2.1.1 release, we'll consider this issue resolved and close the issue. If the problem is not resolved, please update this issue with the relevant details.

**Comment by mattclay** [comment]
Closing as resolved with the 2.1.1 release.
